lib/go-tc/servers.go

```diff
@@ -43,12 +47,113 @@ type ServersV1Response struct {
+	Response []ServerDetailV11 `json:"response"`
+	Alerts
+}
+
+// ServersV3DetailResponse is the JSON object returned for a single server for v3.
+type ServersV3DetailResponse struct {
+	Response []ServerDetailV30 `json:"response"`
 	Alerts
 }
 
+// ServerIpAddress is the data associated with a server's interface's IP address.
+type ServerIpAddress struct {
+	Address        string  `json:"address" db:"address"`
+	Gateway        *string `json:"gateway" db:"gateway"`
+	Interface      string  `json:"interface,omitempty" db:"interface"`
+	Server         uint64  `json:"server,omitempty" db:"server"`
+	ServiceAddress bool    `json:"serviceAddress" db:"service_address"`
+}
+
+// ServerInterfaceInfo is the data associated with a server's interface.
+type ServerInterfaceInfo struct {
+	IpAddresses  []ServerIpAddress `json:"ipAddresses" db:"ipAddresses"`
+	MaxBandwidth *int64            `json:"maxBandwidth" db:"max_bandwidth"`
+	Monitor      bool              `json:"monitor" db:"monitor"`
+	MTU          *uint64           `json:"mtu" db:"mtu"`
+	Name         string            `json:"name" db:"name"`
+}
+
+// Value implements the database/sql/driver.Valuer interface.
+// marshals struct to json to pass back as a json.RawMessage
+func (sii *ServerInterfaceInfo) Value() (driver.Value, error) {
+	b, err := json.Marshal(sii)
+	return b, err
+}
+
+// Scan implements the sql.Scanner interface
+// expects json.RawMessage and unmarshals to a deliveryservice struct
+func (sii *ServerInterfaceInfo) Scan(src interface{}) error {
+	b, ok := src.([]byte)
+	if !ok {
+		return fmt.Errorf("expected deliveryservice in byte array form; got %T", src)
+	}
+
+	return json.Unmarshal([]byte(b), sii)
+}
+
+// LegacyInterfaceDetails is the details for interfaces on servers for API v1 and v2.
+type LegacyInterfaceDetails struct {
+	InterfaceMtu  *int    `json:"interfaceMtu" db:"interface_mtu"`
+	InterfaceName *string `json:"interfaceName" db:"interface_name"`
+	IP6Address    *string `json:"ip6Address" db:"ip6_address"`
+	IP6Gateway    *string `json:"ip6Gateway" db:"ip6_gateway"`
+	IPAddress     *string `json:"ipAddress" db:"ip_address"`
+	IPGateway     *string `json:"ipGateway" db:"ip_gateway"`
+	IPNetmask     *string `json:"ipNetmask" db:"ip_netmask"`
+}
+
+// InterfaceInfoToLegacyInterfaces converts a ServerInterfaceInfo to an
+// equivalent LegacyInterfaceDetails structure. It does this by creating the
+// IP address fields using the "service" interface's IP addresses. All others
+// are discarded, as the legacy format is incapable of representing them.
+func InterfaceInfoToLegacyInterfaces(serverInterfaces []ServerInterfaceInfo) (LegacyInterfaceDetails, error) {
+	var legacyDetails LegacyInterfaceDetails
+
+	for _, intFace := range serverInterfaces {
+		if intFace.MTU != nil {
+			legacyDetails.InterfaceMtu = util.IntPtr(int(*intFace.MTU))
+		}
+
+		legacyDetails.InterfaceName = &intFace.Name
+
+		for _, addr := range intFace.IpAddresses {
+			address := addr.Address
```

Comment:
The returned IP addresses should only be the server's service addresses, otherwise if operators migrate their management interfaces into the new system they will run the risk of occasionally returning management interface details as if they were the actual service IPs of the server.

My implementation uses a check right here and skips the IP if it is not a service IP.